### DISCUSSION TITLE
How to GROUP BY?

### DISCUSSION THREAD

**Discussion by johannbuscail** [supabase/supabase#19517] [Questions] [answered]
Hello, I was wondering if there was a way with `supabase js` to do a `group by` in the `sql` query without using an rpc ?

Example of how i would imagine it:
```ts
const { data: projects, error } = await supabase
		.from("landing_pages")
		.select(`project_id, projects (id, name, description)`)
		.group('project_id');
```

**Comment by GaryAustin1** [comment]
No.  You need to use a view or rpc call to a postgres function and return what you want.

**Comment by steve-chavez** [comment]
PostgREST v12 supports [aggregate functions](https://postgrest.org/en/v12.0/references/api/aggregate_functions.html), this is not yet supported on Supabase but there will be an announcement next week.

**Reply by TomasSestak** [reply]
Is this available already?

**Reply by GaryAustin1** [reply]
If you install the latest database version from infrastructure.

**Reply by jnorris441** [reply]
Error from PostgREST 12 on Supabase: "Use of aggregate functions is not allowed"

Hopefully they will enable aggregate functions soon

**Reply by steve-chavez** [reply, marked as answer]
It needs to be enabled manually.

```sql
ALTER ROLE authenticator SET pgrst.db_aggregates_enabled = 'true';

NOTIFY pgrst, 'reload config';
```

See https://postgrest.org/en/v12.0/references/configuration.html#db-aggregates-enabled

**Reply by sethtjf** [reply]
For anyone planning to do this, I kept seeing "Use of aggregate functions is not allowed" even after running this command. Restarting my instance fixed it.

**Reply by steve-chavez** [reply]
> Restarting my instance fixed it.

Restarting shouldn't be necessary. There was a missing statement in the above snippet, just updated it to include  `NOTIFY pgrst, 'reload config'`.

**Reply by wyozi** [reply]
Is there a way to enable them only for the service role? Trying to enable them on `service_role` only gave me the same error

**Reply by stefan-girlich** [reply]
What is the best way to include this in my codebase? Is it `seed.sql`?

**Reply by seanc** [reply]
How do you group by a property like @johannbuscail was attempting to do with the `project_id` property?

**Reply by MontassarSN** [reply]
after doing that im getting Property 'group' does not exist on type 'PostgrestFilterBuilder

**Reply by GaryAustin1** [reply]
There is no .group even with aggregate functions.
https://supabase.com/blog/postgrest-aggregate-functions
![image](https://github.com/user-attachments/assets/37e61fe3-37c3-4cd4-9d45-393e4bc64b19)

**Reply by veloware** [reply]
@GaryAustin1 without an explicit 'group by' does this not mean that you are restricted to using only grouping columns in your select? and some of those columns you may not want to group by?

Im running into this with an example I have with `invoices` and `invoices_items` (one to many), and I'm trying to `sum` the `invoice_items` `total` for every `invoice` and display it, for that i've followed the docs to invert the embedded resource and use spread - 

```
.from("invoice_items")
    .select(
      `             
        id,
        sum:total.sum(),
        ...invoices(   
          id,      
          created_date,
          due_date,
          projects(id, name, customers(id, org_name)),
          invoice_statuses(id, invoice_id, tenant_id, status, valid_from, valid_to)          
        )
    `,
    )
```

but my assumption here is that will try and group by `invoices.id`, `invoices.created_date`, `invoices.due_date` etc... (and not sure what would happen with the further embedded resources?!)  but I think I only need to group by `invoices.id` ? 

(apologies, my SQL is rusty, but thats my assumption?)

this is so I can show a table like - 

```
invoice001 | 17-09-2024 | €1000
invoice002 | 17-09-2024 | €1500
```

but with the implicit grouping, im getting 3 instances of the first `invoice` as there are 3 `invoice_items`, like - 

```
invoice001 | 17-09-2024 | €1000
invoice001 | 17-09-2024 | €1000
invoice001 | 17-09-2024 | €1000
invoice002 | 17-09-2024 | €1500
```

my assumption is that if I were to be able to group by only `invoices.id` it would return the data as I expected?

**Reply by kyleholgate** [reply]
Some more documentation and examples in the supabase docs would be very helpful for these new aggregate features as it feels unintuitive. 

For example, I just spent longer than I'd like to admit getting a basic count to work because I apparently need to do `.select("status, count()")` and not `.select("status, count(1)")` or `.select("status, count(*)")`. 

I also don't understand the difference between this, which works, and the highly upvoted answer from a maintainer here - https://stackoverflow.com/questions/65612167/how-to-get-count-in-supabase

When would we ever need to do `{ count: 'exact'}`? Does exact relate to pagination?

**Reply by steve-chavez** [reply]
> When would we ever need to do { count: 'exact'}? Does exact relate to pagination?

That's used for the count of the root table. It's special because COUNT is a slow operation on postgres, that's why there's also `count: planned` - which is always fast but can be imprecise.

> and not .select("status, count(1)") or .select("status, count(*)").

Those are invalid syntax, note that the JS library doesn't accept SQL but a REST query grammar. See https://docs.postgrest.org/en/v12/references/api/aggregate_functions.html#the-case-of-count.

**Reply by pirj** [reply]
> Is there a way to enable them only for the service role?

I second this question. I run some queries from an edge function, using the "service role".
My aim with this is twofold:
 - restrict the JS API from using aggregates, eg `id.max()` for no good reason
 - prevent the JS API from high-cost statements (without resorting to [pg_plan_filter_module](https://github.com/pgexperts/pg_plan_filter))

> Trying to enable them on `service_role` only gave me the same error

Same for me.


> What is the best way to include this in my codebase? Is it seed.sql?

This has not been answered, too. I believe the right answer is not using the `seed.sql`, as it is dumped from the linked production database to local, and will be overwritten. There is a `roles.sql` (I could not find it documented anywhere!), which runs locally during `supabase db reset`, and is capable of adjusting roles in your local DB.
I believe, the right approach is to run it in the [web SQL console](https://supabase.com/dashboard/project/_/sql/new).
The `ALTER ROLE authenticator` statement is **not** dumped with `supabase db dump --data-only`, or `supabase db dump`, so, for prod-dev parity, you have to manually add it to `roles.sql`.


> One strategy for preventing potential performance issues is using the [pg_plan_filter_module](https://github.com/pgexperts/pg_plan_filter)

The same question, "where to include this in my codebase"/"how to configure this".
The [doc for pg_plan_filter](https://supabase.com/docs/guides/database/extensions/pg_plan_filter#enable-the-extension) suggests enabling it per connection. But this is pointless, as JS API users may deliberately clamp this.
One option is to turn it for the whole DB `alter database <db-name> set session_preload_libraries = 'plan_filter';`
